src/login/logind-session.c

```diff
@@ -697,6 +699,71 @@ static int session_start_scope(Session *s, sd_bus_message *properties, sd_bus_er
         return 0;
 }
 
+static int session_dispatch_stop_on_idle(sd_event_source *source, uint64_t t, void *userdata) {
+        Session *s = userdata;
+        dual_timestamp ts;
+        usec_t now;
+        int r, idle;
+
+        assert(s);
+
+        if (s->stopping)
+                return 0;
+
+        r = sd_event_now(s->manager->event, CLOCK_MONOTONIC, &now);
+        if (r < 0)
+                return r;
+
+        idle = session_get_idle_hint(s, &ts);
+        if (idle) {
+                log_debug("Session \"%s\" of user \"%s\" is idle, stopping.", s->id, s->user->user_record->user_name);
+
+                r = session_stop(s, /* force */ true);
+                if (r < 0)
+                        log_error_errno(r, "Failed to stop idle session \"%s\": %m", s->id);
+                return r;
+        }
+
+        r = sd_event_source_set_time(source,  ts.monotonic + s->manager->stop_idle_session_usec);
+        if (r < 0)
+                return log_error_errno(r, "Failed to configure stop on idle session event source: %m");
+
+        r = sd_event_source_set_enabled(source, SD_EVENT_ONESHOT);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable stop on idle session event source: %m");
+
+        return 1;
+}
+
+static int session_setup_stop_on_idle_timer(Session *s) {
+        int r;
+
+        assert(s);
+
+        if (s->manager->stop_idle_session_usec == USEC_INFINITY)
+                return 0;
+
+        r = sd_event_add_time_relative(
+                s->manager->event,
+                &s->stop_on_idle_event_source,
+                CLOCK_MONOTONIC,
+                s->manager->stop_idle_session_usec,
+                1,
+                session_dispatch_stop_on_idle, s);
+        if (r < 0)
+                return log_error_errno(r, "Failed to add stop on idle session event source: %m");
+
+        r = sd_event_source_set_priority(s->stop_on_idle_event_source, SD_EVENT_PRIORITY_IMPORTANT);
```

Comment:
I suspect that feature will be used mostly it in order to be in compliance with various certification requirements. So I reckon that event priority is kind of important. Hence, let's just use the default priority instead of important.